### PR TITLE
introduce REST filled DM Channel ID cache resource

### PR DESCRIPTION
**Pull request by FasterSpeeding**
### Summary
introduce REST filled DM Channel ID cache resource
* Remove caching DM channel objects on user objects in favour of new cache resource

### Checklist

- [x] I have run `nox` and all the pipelines have passed.
- [x] I have made unittests according to the code I have added/modified/deleted.

### CODE REVIEW

File: hikari/impl/rest.py [hunk 1]
```diff
@@ -791,7 +796,13 @@ async def fetch_channel(
         route = routes.GET_CHANNEL.compile(channel=channel)
         response = await self._request(route)
         assert isinstance(response, dict)
-        return self._entity_factory.deserialize_channel(response)
+        result = self._entity_factory.deserialize_channel(response)
+
+        if result.type is channels_.ChannelType.DM and self._cache:
+            assert isinstance(result, channels_.DMChannel)
+            self._cache.set_dm_channel_id(result.recipient.id, result.id)
```
Comment (davfsa) [inline hikari/impl/rest.py:803]:
```suggestion
        if self._cache and isinstance(result, channels_.DMChannel):
            self._cache.set_dm_channel_id(result.recipient.id, result.id)
```

**Review by davfsa — COMMENTED** [review]
Looks good, I'm just not so sure if a REST based cache is the way to go. This leaves a lot of open questions, like, should we also be caching channels, emojis, guilds, etc

**Comment by FasterSpeeding** [comment]
> Looks good, I'm just not so sure if a REST based cache is the way to go. This leaves a lot of open questions, like, should we also be caching channels, emojis, guilds, etc

Except A you're already caching EST data just in a kinda bad way which leads to stale data being returned and doesn't have any formalised interface for it.

B how the cache functions around removing entries is implementation detail and cacheing based on REST would only work for resource implemention which aren't reliant on delete events to remove resources from it (which only covers messages, users and dm channel ids to the standard implementation )

**Comment by davfsa** [comment]
Sounds good then. just the little comment I made with the review

File: hikari/impl/rest.py (hunk 1 above)
Comment (FasterSpeeding) [reply hikari/impl/rest.py:803]: Wouldn't that be slower for optimised runs for what is the exact same result?

**Review by davfsa — APPROVED** [review]

**Pull request merged**